ballista/rust/scheduler/src/state/execution_graph_dot.rs

```diff
@@ -0,0 +1,537 @@
+                let parent_shuffle_write_node = format!("stage_{}_0", parent_stage_id);
+                links.push(format!("{} -> {}", parent_shuffle_write_node, reader_node,));
+            }
+            // keep the order deterministic
+            links.sort();
+            for link in links {
+                writeln!(&mut dot, "\t{}", link)?;
+            }
+        }
+
+        writeln!(&mut dot, "}}")?; // end of digraph
+
+        Ok(dot)
+    }
+}
+
+/// Write the query tree for a single stage and build metadata needed to later draw
+/// the links between the stages
+fn write_stage_plan(
+    f: &mut String,
+    prefix: &str,
+    plan: &Arc<dyn ExecutionPlan>,
+    i: usize,
+) -> Result<StagePlanState, fmt::Error> {
+    let mut state = StagePlanState {
+        readers: HashMap::new(),
+    };
+    write_plan_recursive(f, prefix, plan, i, &mut state)?;
+    Ok(state)
+}
+
+fn write_plan_recursive(
+    f: &mut String,
+    prefix: &str,
+    plan: &Arc<dyn ExecutionPlan>,
+    i: usize,
+    state: &mut StagePlanState,
+) -> Result<(), fmt::Error> {
+    let node_name = format!("{}_{}", prefix, i);
+    let display_name = get_operator_name(plan);
+
+    if let Some(reader) = plan.as_any().downcast_ref::<ShuffleReaderExec>() {
+        for part in &reader.partition {
+            for loc in part {
+                state
+                    .readers
+                    .insert(node_name.clone(), loc.partition_id.stage_id);
+            }
+        }
+    } else if let Some(reader) = plan.as_any().downcast_ref::<UnresolvedShuffleExec>() {
+        state.readers.insert(node_name.clone(), reader.stage_id);
+    }
+
+    let mut metrics_str = vec![];
+    if let Some(metrics) = plan.metrics() {
+        if let Some(x) = metrics.output_rows() {
+            metrics_str.push(format!("output_rows={}", x))
+        }
+        if let Some(x) = metrics.elapsed_compute() {
+            metrics_str.push(format!("elapsed_compute={}", x))
+        }
+    }
+    if metrics_str.is_empty() {
+        writeln!(
+            f,
+            "\t\t{} [shape=box, label=\"{}\"]",
+            node_name, display_name
+        )?;
+    } else {
+        writeln!(
+            f,
+            "\t\t{} [shape=box, label=\"{}
+{}\"]",
+            node_name,
+            display_name,
+            metrics_str.join(", ")
+        )?;
+    }
+
+    for (j, child) in plan.children().into_iter().enumerate() {
+        write_plan_recursive(f, &node_name, &child, j, state)?;
+        // write link from child to parent
+        writeln!(f, "\t\t{}_{} -> {}", node_name, j, node_name)?;
+    }
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct StagePlanState {
+    /// map from reader node name to parent stage id
+    readers: HashMap<String, usize>,
+}
+
+/// Make strings dot-friendly
+fn sanitize_dot_label(str: &str) -> String {
+    // TODO make max length configurable eventually
+    sanitize(str, Some(100))
+}
+
+/// Make strings dot-friendly
```

Comment:
AFAIK labels can be quoted like this 

`"Thïs įs a 🦄 and a \" quote"`

Reply:
So surround by quotes and escape quote character is all that's needed.

Reply:
I did try this, and it failed, but maybe I made a mistake ... I will try again in the next PR